---
title: "Figure 2"
author: "A. M. Chakrabarti"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_notebook: 
    code_folding: hide
    toc: yes
    toc_depth: 2
    toc_float: yes
    number_sections: yes
    theme: flatly
    highlight: haddock
    fig_align: center
---

# Setup

```{r}
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(tictoc))
suppressPackageStartupMessages(library(toscatools))
suppressPackageStartupMessages(library(rtracklayer))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(scales))
suppressPackageStartupMessages(library(ggthemes))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(UpSetR))
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(patchwork))
```

```{r}
results.path <- "/Volumes/lab-luscomben/home/shared/projects/ira-nobby/comp_hiclip/revisions"
plot.path <- "~/Dropbox (The Francis Crick)/comp_hiclip/revisions/plots/figure_2"
if(!dir.exists(plot.path)) dir.create(plot.path, recursive = TRUE)

ht_colours <- c("Linker" = "#8175aa", "Direct" = "#6fb899", "Short-range" = "#31a1b3")
```

# Load data

```{r}
# linker.dt <- rbindlist(list(fread(file.path(results.path, "results_linker/stau1_high.hybrids.gc.annotated.tsv.gz"))[, sample := "stau1_high"],
#                             fread(file.path(results.path, "results_linker/stau1_low.hybrids.gc.annotated.tsv.gz"))[, sample := "stau1_low"]),
#                        use.names = TRUE)

linker.dt <- fread(file.path(results.path, "results_linker/all.atlas.clustered.tsv.gz"))
nolinker.dt <- fread(file.path(results.path, "results_nolinker/atlas/all.atlas.clustered.tsv.gz"))
nolinker.dt[, `:=` (cluster_hybrid_count = NULL,
                    cluster_hybrid_count.x = NULL,
                    sample = tstrsplit(sample, "\\.")[[1]])]
setnames(nolinker.dt, "cluster_hybrid_count.y", "cluster_hybrid_count")

hybrids.dt <- rbindlist(list(linker.dt, nolinker.dt),
                        use.names = TRUE, 
                        fill = TRUE)

# Adjust parameter names for plots
hybrids.dt[, exp := ifelse(orientation == "linker", "Linker", "Direct")]
hybrids.dt[, type := ifelse(type == "intragenic", "intra-transcript", "inter-transcript")]
hybrids.dt[sample == "stau1_high", sample := "High RNase"]
hybrids.dt[sample == "stau1_low", sample := "Low RNase"]
```

# Figure 2 A - Counts

```{r}
p.A <- ggplot(hybrids.dt[, .N, by = .(exp, sample)], aes(x = exp, y = N, fill = sample)) +
  geom_col(width = 0.5) +
  coord_flip() + 
  scale_x_discrete(limits = c("Direct", "Linker")) +
  scale_y_continuous(breaks = seq(0, 5e4, 1e4)) +
  scale_fill_tableau() +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "",
       y = "Number of hybrids",
       fill = "")

ggsave(p.A, filename = file.path(plot.path, "linker_nolinker_counts.pdf"), width = 7, height = 4)

p.A
```

There are `r nrow(nolinker.dt)/nrow(linker.dt)` times as many hybrids recovered without a linker

# Figure 2 B - Intra-inter

```{r}
p.B <- hybrids.dt %>% 
  group_by(exp, type) %>% 
  summarise(reads = n()) %>%
  mutate(percentage = reads/sum(reads)) %>%
  ggplot(aes(x = exp, y = percentage, fill = type)) +
  geom_col(width = 0.5, position = "fill") +
  geom_text(aes(label = paste0(round(percentage * 100,1),"%")), 
            position = position_stack(vjust=0.5),
            colour = "white",
            size = 5) +
  coord_flip() + scale_x_discrete(limits = c("Direct", "Linker")) +
  scale_y_continuous(label = percent) +
  scale_fill_tableau(palette = "Seattle Grays", guide = guide_legend(reverse = TRUE)) +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "",
       y = "Percentage of hybrids",
       fill = "")

ggsave(p.B, filename = file.path(plot.path, "linker_nolinker_interintra.pdf"), width = 7, height = 4)

p.B
```

### Inter-transcript

```{r}
inter.dt <- hybrids.dt[type == "inter-transcript"]
# inter.dt <- inter.dt[, .N, by = .(L_biotype, R_biotype)]
inter.dt[, `:=` (biotype.x = ifelse(L_biotype <= R_biotype, L_biotype, R_biotype),
                 biotype.y = ifelse(L_biotype <= R_biotype, R_biotype, L_biotype))]
inter.dt[, link := paste0(biotype.y, "-", biotype.x)]

inter.dt[grepl("rRNA", biotype.y), simple.biotype.y := "rRNA"]
inter.dt[grepl("rRNA", biotype.x), simple.biotype.x := "rRNA"]
inter.dt[grepl("mRNA", biotype.y), simple.biotype.y := "mRNA"]
inter.dt[grepl("mRNA", biotype.x), simple.biotype.x := "mRNA"]
inter.dt[grepl("lncRNA", biotype.y), simple.biotype.y := "lncRNA"]
inter.dt[grepl("lncRNA", biotype.x), simple.biotype.x := "lncRNA"]
inter.dt[is.na(simple.biotype.y), simple.biotype.y := biotype.y]
inter.dt[is.na(simple.biotype.x), simple.biotype.x := biotype.x]
inter.dt[, simple.link := paste0(simple.biotype.y, "-", simple.biotype.x)]

inter.summarised.count.dt <- inter.dt[, .N, by = .(simple.link)]
setorder(inter.summarised.count.dt, -N)

inter.count.dt <- inter.dt[, .N, by = .(simple.link, exp)]
p.inter <- ggplot(inter.count.dt[simple.link %in% inter.summarised.count.dt[1:5]$simple.link], aes(x = reorder(simple.link, N), y = N, fill = exp)) +
  geom_col(position = "dodge") +
  coord_flip() +
  # facet_grid(. ~ exp) +
  scale_fill_manual(values = ht_colours) +
  theme_minimal_grid() + theme(legend.position = "bottom") +
  labs(y = "Number of hybrids",
       x = "",
       fill = "")

ggsave(p.inter, filename = file.path(plot.path, "linker_nolinker_inter_top5.pdf"), width = 7, height = 4)

p.inter
```

### mRNA-mRNA clusters

```{r}
# inter.nolinker.clusters <- collapse_clusters(nolinker.dt[L_seqnames != R_seqnames][!L_biotype %in% c("rRNA", "tRNA")][!R_biotype %in% c("rRNA", "tRNA")][cluster_hybrid_count >= 2], mode = "median")
inter.nolinker.clusters <- collapse_clusters(nolinker.dt[L_seqnames != R_seqnames][grepl("mRNA", L_biotype)][grepl("mRNA", R_biotype)][cluster_hybrid_count >= 2], mode = "median")
inter.nolinker.clusters <- convert_coordinates(inter.nolinker.clusters, 
                                               genes.gr = import.gff2("/Volumes/lab-luscomben/home/shared/projects/ira-nobby/comp_hiclip/ref/GRCh38.gencode_v33.tx.gtf.gz"))

inter.nolinker.clusters <- annotate_hybrids(inter.nolinker.clusters, 
                                            regions.gr = import.gff2("/Volumes/lab-luscomben/home/shared/projects/ira-nobby/comp_hiclip/ref/icount_mini_utr3/regions.gtf.gz"))

inter.nolinker.clusters[, L_region := factor(L_region, levels = c("rRNA", "tRNA", "UTR5", "CDS", "intron", "UTR3", "ncRNA"))]
inter.nolinker.clusters[, R_region := factor(R_region, levels = c("rRNA", "tRNA", "UTR5", "CDS", "intron", "UTR3", "ncRNA"))]

inter.nolinker.clusters[, `:=` (region.x = ifelse(as.integer(L_region) <= as.integer(R_region), as.character(L_region), as.character(R_region)),
                      region.y = ifelse(as.integer(L_region) <= as.integer(R_region), as.character(R_region), as.character(L_region)))]
inter.nolinker.clusters[, mrna.link := paste0(region.x, "-", region.y)]
inter.nolinker.clusters.count.dt <- inter.nolinker.clusters[, .N, by = mrna.link]
setorder(inter.nolinker.clusters.count.dt, -N)

p.intermrna <- ggplot(inter.nolinker.clusters.count.dt, aes(x = reorder(mrna.link, N), y = N)) +
  geom_col() +
  coord_flip() +
  theme_minimal_grid() + 
  labs(y = "Number of duplexes",
       x = "")

ggsave(p.intermrna, filename = file.path(plot.path, "linker_nolinker_inter_mrna_clusters.pdf"), width = 4, height = 4)

p.intermrna
```

# Figure 2 C - Regions

```{r}
arms.dt <- rbindlist(list(hybrids.dt[, .(exp, L_region)], hybrids.dt[, .(exp, R_region)]), use.names = FALSE)
arms.dt$L_region <- factor(arms.dt$L_region, levels = rev(c("rRNA", "tRNA", "UTR5", "CDS", "intron", "UTR3", "ncRNA")))

p.C <- arms.dt %>%
  group_by(L_region, exp) %>%
  summarise(arms = n()) %>%
  group_by(exp) %>%
  mutate(percentage = arms/sum(arms)) %>%
  ggplot(aes(x = exp, y = percentage, fill = L_region)) +
  geom_col(width = 0.5, position = "fill") +
  geom_text(aes(label = ifelse(percentage > 0.05, paste0(round(percentage * 100, 1), "%"), "")), 
            position = position_stack(vjust=0.5),
            colour = "white",
            size = 5) +
  coord_flip() +
  scale_x_discrete(limits = c("Direct", "Linker")) +
  scale_y_continuous(label = percent) +
  scale_fill_tableau(palette = "Miller Stone", 
                     direction = -1, 
                     guide = guide_legend(reverse = TRUE)) +
  theme_minimal_grid() + 
  theme(legend.position = "top") +
  labs(x = "",
       y = "Percentage of hybrids",
       fill = "")

ggsave(p.C, filename = file.path(plot.path, "linker_nolinker_regions.pdf"), width = 7, height = 4)

p.C
```

```{r}
round(prop.table(table(arms.dt[exp == "Linker"]$L_region)) * 100, 1)
round(prop.table(table(arms.dt[exp == "Direct"]$L_region)) * 100, 1)
```

# Figure 2 D - Overlaps

```{r}
linker.clusters <- collapse_clusters(linker.dt[L_seqnames == R_seqnames][!L_biotype %in% c("rRNA", "tRNA")][cluster_hybrid_count >= 2], mode = "median")
nolinker.clusters <- collapse_clusters(nolinker.dt[L_seqnames == R_seqnames][!L_biotype %in% c("rRNA", "tRNA")][cluster_hybrid_count >= 2], mode = "median")

# linker.clusters <- collapse_clusters(linker.dt[L_seqnames == R_seqnames][cluster_hybrid_count >= 2], mode = "wide")
# nolinker.clusters <- collapse_clusters(nolinker.dt[L_seqnames == R_seqnames][cluster_hybrid_count >= 2], mode = "wide")

linker.clusters <- reorient_hybrids(linker.clusters)
nolinker.clusters <- reorient_hybrids(nolinker.clusters)

bedpe.colnames <- c("L_seqnames", "L_start", "L_end", "R_seqnames", "R_start", "R_end", "name", "count", "L_strand", "R_strand")

linker.bedpe.dt <- linker.clusters[, ..bedpe.colnames]
linker.bedpe.dt[, `:=`(
  L_start = L_start - 1,
  R_start = R_start - 1
)]
nolinker.bedpe.dt <- nolinker.clusters[, ..bedpe.colnames]
nolinker.bedpe.dt[, `:=`(
  L_start = L_start - 1,
  R_start = R_start - 1
)]

linker.bedpe <- tempfile(tmpdir = getwd(), fileext = ".bedpe")
nolinker.bedpe <- tempfile(tmpdir = getwd(), fileext = ".bedpe")
ol <- tempfile(tmpdir = getwd(), fileext = ".bedpe")

fwrite(linker.bedpe.dt, file = linker.bedpe, sep = "\t", col.names = FALSE)
fwrite(nolinker.bedpe.dt, file = nolinker.bedpe, sep = "\t", col.names = FALSE)

cmd <- paste("bedtools pairtopair -rdn -a", linker.bedpe, "-b", nolinker.bedpe, ">", ol)
system(cmd)
bedpe.dt <- fread(ol, col.names = c(paste0(bedpe.colnames, ".x"), paste0(bedpe.colnames, ".y")))

invisible(file.remove(linker.bedpe))
invisible(file.remove(nolinker.bedpe))
invisible(file.remove(ol))

table(duplicated(bedpe.dt$name.x))
table(duplicated(bedpe.dt$name.y))

stopifnot(sum(linker.clusters$name %in% bedpe.dt$name.x) == sum(nolinker.clusters$name %in% bedpe.dt$name.y))
ol <- sum(linker.clusters$name %in% bedpe.dt$name.x)

upset.input <- c("Linker" = nrow(linker.clusters) - ol,
               "Direct" = nrow(nolinker.clusters) - ol,
               "Linker&Direct" = ol)

pdf(file.path(plot.path, "linker_nolinker_upset.pdf"), width = 7, height = 5)
upset(fromExpression(upset.input), 
      order.by = "freq",
      mainbar.y.label = "Intersecting number of duplexes",
      sets.x.label = "Total number of duplexes")
dev.off()

upset(fromExpression(upset.input), 
      order.by = "freq",
      mainbar.y.label = "Intersecting number of duplexes",
      sets.x.label = "Total number of duplexes")
```

## Figure 2 D - Inset - RNA abundance of overlapping genes

```{r}
ol.genes <- unique(c(bedpe.dt$L_seqnames.x, bedpe.dt$L_seqnames.y, bedpe.dt$R_seqnames.x, bedpe.dt$R_seqnames.y))
nol.genes <- unique(c(linker.bedpe.dt$L_seqnames, linker.bedpe.dt$R_seqnames, nolinker.bedpe.dt$L_seqnames, nolinker.bedpe.dt$R_seqnames))

gene.counts.dt <- fread("~/Dropbox (The Francis Crick)/comp_hiclip/4suseq/GSE84722_ST1.txt.gz")
gene.counts.dt$mean_count <- rowMeans(gene.counts.dt[, .(RZ_Mature_copies, RH_Mature_copies, PA_Mature_copies)], na.rm = TRUE)
gene.counts.dt[, ensg := tstrsplit(Gene, "\\.")[[1]]]

counts.dt <- data.table(gene = c(ol.genes, nol.genes),
                        type = c(rep("Overlap", length(ol.genes)), rep("No overlap", length(nol.genes))))
counts.dt <- counts.dt[!grepl("_", gene)]
counts.dt[, ensg := tstrsplit(tstrsplit(gene, ":")[[2]], "\\.")[[1]]]

counts.dt <- merge(counts.dt, gene.counts.dt[, .(ensg, mean_count)], by = "ensg")

p <- ggplot(counts.dt, aes(x = type, y = mean_count)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  scale_y_log10() +
  annotation_logticks(sides = "l") +
  # scale_fill_brewer(palette = "Accent") +
  coord_cartesian(ylim = c(1e-1, 1e3)) +
  # geom_label(data = gene.counts.dt[mean_count > 0, .N, by = .(stau)], aes(label = N, y = 0), vjust = -0.1) +
  labs(x = "",
       y = "Mean gene copy number",
       fill = "") +
  theme_minimal_grid() + theme(legend.position = "none")

ggsave(p, filename = file.path(plot.path, "overlap_rnaabundance.pdf"), width = 4, height = 4)

p
```

# Figure 2 E - hybridisation energy

```{r}
mfe.dt <- rbindlist(list(nolinker.dt[L_seqnames == R_seqnames][!L_biotype %in% c("rRNA", "tRNA")][grep("^C", cluster)][, mean(mean_shuffled_mfe), by = .(L_seqnames, cluster)][, `:=` (exp = "Direct", mfe = "Shuffled")],
                         nolinker.dt[L_seqnames == R_seqnames][!L_biotype %in% c("rRNA", "tRNA")][grep("^C", cluster)][, mean(mfe), by = .(L_seqnames, cluster)][, `:=` (exp = "Direct", mfe = "Duplex")],
                         linker.dt[L_seqnames == R_seqnames][!L_biotype %in% c("rRNA", "tRNA")][grep("^C", cluster)][, mean(mean_shuffled_mfe), by = .(L_seqnames, cluster)][, `:=` (exp = "Linker", mfe = "Shuffled")],
                         linker.dt[L_seqnames == R_seqnames][!L_biotype %in% c("rRNA", "tRNA")][grep("^C", cluster)][, mean(mfe), by = .(L_seqnames, cluster)][, `:=` (exp = "Linker", mfe = "Duplex")]))

setnames(mfe.dt, "V1", "energy")

p.E <- ggplot(mfe.dt, aes(x = energy, colour = exp, linetype = mfe)) +
  geom_density(size = 1) +
  scale_colour_manual(values = ht_colours) +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "Hybridisation energy",
       y = "Density",
       colour = "",
       linetype = "")  

ggsave(p.E, filename = file.path(plot.path, "linker_nolinker_mfe.pdf"), width = 7, height = 4)

p.E
```

# Figure 2 F - Duplex span

```{r}
linker.clusters <- collapse_clusters(linker.dt[L_seqnames == R_seqnames][!L_biotype %in% c("rRNA", "tRNA")][cluster_hybrid_count >= 2], mode = "median")
nolinker.clusters <- collapse_clusters(nolinker.dt[L_seqnames == R_seqnames][!L_biotype %in% c("rRNA", "tRNA")][cluster_hybrid_count >= 2], mode = "median")

nolinker.clusters[, loop := R_start - L_end - 1][, exp := "Linker"]
linker.clusters[, loop := R_start - L_end - 1][, exp := "Direct"]

intragenic.clusters.dt <- rbindlist(list(nolinker.clusters, linker.clusters))

p.F <- ggplot(intragenic.clusters.dt, aes(x = loop, colour = exp)) +
  geom_density(size = 1) +
  scale_colour_manual(values = ht_colours) +
  scale_x_log10(label = comma) + annotation_logticks(sides = "b") +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "Duplex span",
       y = "Density",
       colour = "",
       linetype = "")  

ggsave(p.F, filename = file.path(plot.path, "intragenic_linker_nolinker_duplex_span.pdf"), width = 7, height = 4)

p.F
```
