---
title: "Figure 6"
author: "A. M. Chakrabarti"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_notebook: 
    code_folding: hide
    toc: yes
    toc_depth: 2
    toc_float: yes
    number_sections: yes
    theme: flatly
    highlight: haddock
    fig_align: center
---

```{r, warning = FALSE}
suppressPackageStartupMessages(library(toscatools))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(scales))
suppressPackageStartupMessages(library(ggthemes))
suppressPackageStartupMessages(library(UpSetR))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(GenomicFeatures))
suppressPackageStartupMessages(library(rtracklayer))
suppressPackageStartupMessages(library(parallel))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(patchwork))
suppressPackageStartupMessages(library(dunn.test))
suppressPackageStartupMessages(library(tictoc))
set.seed(42)
```

```{r}
save_plots <- TRUE
results.path <- "/Volumes/lab-luscomben/home/shared/projects/ira-nobby/comp_hiclip/revisions"
plot.path <- "~/Dropbox (The Francis Crick)/comp_hiclip/revisions/plots/figure_6"
if(!dir.exists(plot.path)) dir.create(plot.path)

ht_colours <- c("Linker" = "#8175aa", "Direct" = "#6fb899", "Short-range" = "#31a1b3") # Nuriel Stone
gt_colours <- c("None" = "#767f8b", "None\n(Expression\nMatched)" = "#767f8b", "PARIS" = "#f28e2b", "STAU1" = "#4e79a7")
metabolism_colours = c(C = "#414451", B = "#60636a", A = "#a5acaf")
```

# Load atlas

```{r}
# duplexes.dt <- fread("~/Dropbox (The Francis Crick)/comp_hiclip/stau1_atlas/merged_atlas.clusters.collapsed_plus_nonhybrids.tsv.gz")
# original.utr3.duplexes.dt <- duplexes.dt[L_seqnames == R_seqnames][L_region == "UTR3"][R_region == "UTR3"]
# original.utr3.duplexes.dt <- original.utr3.duplexes.dt[!grep("^ID", cluster)]

duplexes.dt <- fread(file.path(results.path, "atlas/merged.atlas.clusters.tsv.gz"))
utr3.duplexes.dt <- duplexes.dt[L_seqnames == R_seqnames][L_region == "UTR3"][R_region == "UTR3"]
utr3.duplexes.dt <- utr3.duplexes.dt[!grep("^ID", cluster)]

paris.dt <- fread(file.path(results.path, "paris/results_paris/atlas_clusters/all.atlas_clusters.gc.annotated.tsv.gz"))
utr3.paris.dt <- paris.dt[L_seqnames == R_seqnames][L_region == "UTR3"][R_region == "UTR3"] # To check all are intra 3' UTR
```

## Categorise genes

```{r}
utr3.duplexes.dt[, paris := ifelse(L_gene_id %in% unique(utr3.paris.dt$L_gene_id), TRUE, FALSE)]
stau1.genes <- unique(tstrsplit(utr3.duplexes.dt[paris == FALSE]$L_gene_id, "\\.")[[1]])
stau1_paris.genes <- unique(tstrsplit(utr3.duplexes.dt[paris == TRUE]$L_gene_id, "\\.")[[1]])
paris.genes <- unique(tstrsplit(utr3.paris.dt$L_gene_id, "\\.")[[1]])
paris.genes <- paris.genes[!paris.genes %in% c(stau1.genes, stau1_paris.genes)]
```

For our downstream analysis, we keep 3' UTR intramolecular duplexes. This leaves `r nrow(utr3.duplexes.dt)` STAU1 duplexes across `r length(unique(utr3.duplexes.dt$L_gene_id))` genes and `r nrow(utr3.paris.dt)` PARIS duplexes across `r length(unique(utr3.paris.dt$L_gene_id))` genes. 

## Calculate loops and introns

```{r}
tic()
utr3.duplexes.dt[, ensg := sapply(strsplit(L_gene_id, "\\."), "[[", 1)]

TxDb <- loadDb("/Volumes/lab-luscomben/home/shared/projects/ira-nobby/comp_hiclip/ref/gencode.v33.txdb.sqlite")
# TxDb <- loadDb("~/Dropbox (The Francis Crick)/comp_hiclip/revisions/camp/gencode.v33.txdb.sqlite")

txlengths.dt <- data.table(transcriptLengths(TxDb, with.cds_len = TRUE,
                                             with.utr5_len = TRUE,
                                             with.utr3_len = TRUE))
setorder(txlengths.dt, gene_id, -utr3_len)

utr3.grl <- threeUTRsByTranscript(TxDb, use.names = TRUE)

L.gr <- convert_to_granges(utr3.duplexes.dt, arm = "L", genomic = TRUE)
R.gr <- convert_to_granges(utr3.duplexes.dt, arm = "R", genomic = TRUE)
stopifnot(all(countOverlaps(L.gr, utr3.grl) != 0))
stopifnot(all(countOverlaps(R.gr, utr3.grl) != 0))
duplexes.grl <- split(c(L.gr, R.gr), c(L.gr, R.gr)$name)

utr3.dt <- mclapply(1:length(duplexes.grl), function(i) {
  
# utr3.dt <- lapply(1:length(duplexes.grl), function(i) {
  
  # message(i)
  x <- duplexes.grl[[i]]
  
  # First get 3' UTR for the transcript
  ol <- findOverlaps(x, utr3.grl)
  stopifnot(length(ol) > 0)
  
  if(length(unique(subjectHits(ol))) == 1) {
  
    utr3 <- utr3.grl[unique(subjectHits(ol))]
    
    utr3 <- unlist(utr3)
    utr3$tx_name <- names(utr3)
    names(utr3) <- NULL
    
  } else {
    
    sel.ol <- unique(subjectHits(ol))
    sel.ol <- sel.ol[sapply(sel.ol, function(y) all(queryHits(ol[subjectHits(ol) == y]) == c(1, 2)))] # Keep only ones where both arms overlapped by same 3' UTR overlapped
    # stopifnot(length(sel.ol) != 0)
    if(length(sel.ol) == 0) {
      return(NA) # these are intra-genic UTR3-UTR3 but across different transcripts
    }
    
    utr3 <- utr3.grl[sel.ol]
    
    utr3.tx.dt <- txlengths.dt[tx_name %in% names(utr3)] # select matching tx
    utr3.tx.dt[gene_id %in% x$L_gene_id] # keep only same genes
    utr3 <- utr3[names(utr3) %in% utr3.tx.dt[1]] # select longest 3' UTR 
    
    utr3 <- unlist(utr3)
    utr3$tx_name <- names(utr3)
    names(utr3) <- NULL
    
  }
  
  # Then work out introns
  utr3.dt <- data.table(name = unique(x$name),
                        ensg = unique(x$ensg),
                        utr3_start = min(start(utr3)),
                        utr3_end = max(end(utr3)),
                        utr3_width = sum(width(utr3)))
  
  g <- gaps(utr3, start = min(start(utr3)), end = max(end(utr3)))
  g <- g[seqnames(g) == unique(seqnames(utr3))]
  g <- g[strand(g) == unique(strand(utr3))]
  
  # Reversed order if strand is neg to match utr3
  if(unique(strand(utr3)) == "-") {
    g <- sort(g, decreasing = TRUE)
  } else {
    g <- sort(g)
  }

  utr3.dt[, utr3_intron := sum(width(g))]
  
  # Between arms   
  L.ol <- max(which(countOverlaps(utr3, x[1]) > 0)) # if intron in arm, select
  R.ol <- min(which(countOverlaps(utr3, x[2]) > 0))
  
  if(L.ol == R.ol) {
    
    utr3.dt[, intron := 0]
    
  } else {
    
    stopifnot(R.ol > L.ol) # Should be this for all
    k <- seq(L.ol, R.ol - 1, by = 1) # which introns to keep
    utr3.dt[, intron := sum(width(g[k]))]
     
  }
  
  # Relative starts accounting for introns
  if(as.character(strand(x[1])) == "-") {
    
    L_abs_start <- max(end(utr3)) - end(x[1]) + 1
    if(L.ol != 1) {
      L_abs_start <- L_abs_start - sum(width(g[seq(1, L.ol - 1, by = 1)]))
    }
    R_abs_start <- max(end(utr3)) - end(x[2]) + 1
    if(R.ol != 1) {
      R_abs_start <- R_abs_start - sum(width(g[seq(1, R.ol - 1, by = 1)]))
    }
    
  } else {
    
    L_abs_start <- start(x[1]) - min(start(utr3)) + 1
    if(L.ol != 1) {
      L_abs_start <- L_abs_start - sum(width(g[seq(1, L.ol - 1)]))
    }
    R_abs_start <- start(x[2]) - min(start(utr3)) + 1
    if(R.ol != 1) {
    R_abs_start <- R_abs_start - sum(width(g[seq(1, R.ol - 1)]))
    }
    
  }
  
  # Relative ends accounting for introns
  if(as.character(strand(x[2])) == "-") {
    
    R_abs_end <- start(x[2]) - min(start(utr3)) + 1
    if(R.ol != length(utr3)) {
      R_abs_end <- R_abs_end - sum(width(g[seq(R.ol, length(utr3) - 1, by = 1)]))
    }
    
  } else {
    
    R_abs_end <- max(end(utr3)) - end(x[2]) + 1
    if(R.ol != length(utr3)) {
    R_abs_end <- R_abs_end - sum(width(g[seq(R.ol, length(utr3) - 1)]))
    }
    
  }  
  
  utr3.dt[, `:=` (L_abs_start = L_abs_start,
                  R_abs_start = R_abs_start,
                  R_abs_end = R_abs_end)]
  
  stopifnot(all(c(utr3.dt$L_abs_start,
                  utr3.dt$R_abs_start, 
                  utr3.dt$R_abs_end) < utr3.dt$utr3_width))
  
  return(utr3.dt)
  
}, mc.cores = 4)

stopifnot(all(elementNROWS(utr3.dt) == 1))

table(unlist(lapply(utr3.dt, length)) == 10) # 1 is NA, i.e. across different 3' UTRs
utr3.dt <- utr3.dt[unlist(lapply(utr3.dt, length)) == 10]

utr3.dt <- rbindlist(utr3.dt)
utr3.duplexes.dt <- merge(utr3.duplexes.dt, utr3.dt, by = c("name", "ensg"))
utr3.duplexes.dt[R_abs_end < 0, R_abs_end := 0] # Some extend off end of annotated transcript
utr3.duplexes.dt[, R_abs_end := -R_abs_end] # As distance from end

# Calculations
utr3.duplexes.dt[, `:=` (loop = R_start - L_end - 1 - intron,
                         L_rel_start = L_abs_start/utr3_width,
                         R_rel_start = R_abs_start/utr3_width)]
stopifnot(nrow(utr3.duplexes.dt[loop > utr3_width]) == 0)
utr3.duplexes.dt[, circ_score := loop/utr3_width]
toc()
```

# RNA abundance

```{r}
# ==========
# From Neel:
geneInfo <- read.table("~/Dropbox (The Francis Crick)/comp_hiclip/4suseq/replicationAnalysis/geneInfo.txt", header=T)
geneInfo$Annotation <- factor(geneInfo$Annotation, levels = rownames(as.data.frame.array(sort(table(geneInfo$Annotation), decreasing = T))))
rownames(geneInfo) <- geneInfo$Gene
keepType <- rownames(as.data.frame.array(sort(table(geneInfo$Annotation), decreasing = T)))[c(1:4,9,10,13,20)]

geneInfo$Simple <- plyr::mapvalues(geneInfo$Annotation, from = keepType, to = c("protein_coding","pseudogene","lncRNA","lncRNA","lncRNA","lncRNA","lncRNA","lncRNA"))
geneInfo <- subset(geneInfo, Annotation %in% keepType)
geneInfo <- droplevels(geneInfo)
# ==========

# Now integrate
gene.info.dt <- data.table(geneInfo)

# Get means
gene.counts.dt <- fread("~/Dropbox (The Francis Crick)/comp_hiclip/4suseq/GSE84722_ST1.txt.gz")
gene.counts.dt$mean_count <- rowMeans(gene.counts.dt[, .(RZ_Mature_copies, RH_Mature_copies, PA_Mature_copies)], na.rm = TRUE)

# Select protein coding genes
gene.counts.dt <- gene.counts.dt[Gene %in% gene.info.dt[Simple == "protein_coding"]$Gene]

# Annotate with Staufen 3' UTR binding
gene.counts.dt[, ensg := sapply(strsplit(Gene, "\\."), "[[", 1)]
gene.counts.dt[ensg %in% stau1.genes, stau := "STAU1"][ensg %in% stau1_paris.genes, stau := "STAU1"][ensg %in% paris.genes, stau := "PARIS"][is.na(stau), stau := "None"]
gene.counts.dt$stau <- factor(gene.counts.dt$stau, levels = c("None", "PARIS", "STAU1", "STAU1 & PARIS"))

# d <- dunn.test(x = gene.counts.dt[mean_count > 0]$mean_count, 
#                g = gene.counts.dt[mean_count > 0]$stau,
#                method = "bh",
#                alpha = 0.01)

# Plot
p <- ggplot(gene.counts.dt[mean_count > 0], aes(x = stau, y = mean_count)) +
  geom_violin(aes(fill = stau)) +
  geom_boxplot(width = 0.1, outlier.shape = NA) +
  scale_y_log10(label = comma) +
  annotation_logticks(sides = "l") +
  scale_fill_manual(values = gt_colours) +
  geom_text(data = gene.counts.dt[mean_count > 0, .N, by = .(stau)], aes(label = N, y = 0.0001), vjust = -0.1) +
  labs(x = "",
       y = "Mean gene expression",
       fill = "") +
  theme_minimal_grid() + theme(legend.position = "none")

if(save_plots) ggsave(p, filename = file.path(plot.path, "rna_abundance.pdf"), width = 5, height = 4)
p
```
# RNA metabolism

```{r}
# From Neel
# import rates
clean_ddp <- function(file, base) {
  file <- read.table(file, header=T, sep="\t")
  
  file <-file[,grep("_ddp", colnames(file), invert = T)]
  colnames(file) <- gsub("pulse", base, colnames(file))
  colnames(file) <- gsub("min", "", colnames(file))
  assign("tmp", file, envir=.GlobalEnv)
}

rates <- cbind(clean_ddp("~/Dropbox (The Francis Crick)/comp_hiclip/4suseq/replicationAnalysis/synthesis_rates.txt", "Syn_"),
               clean_ddp("~/Dropbox (The Francis Crick)/comp_hiclip/4suseq/replicationAnalysis/processing_rates.txt", "Proc_"),
               clean_ddp("~/Dropbox (The Francis Crick)/comp_hiclip/4suseq/replicationAnalysis/degradation_rates.txt", "Deg_"))

rates <- rates[is.finite(rates$Deg_30),] # Remove deg rates "NA" -> Stefano what are these?

ratesInfo <- data.frame(
  Gene=rownames(rates),
  Syn = rowMeans(rates[,grep("Syn_",colnames(rates))]),
  Proc = rowMeans(rates[,grep("Proc_",colnames(rates))]),
  Deg = rowMeans(rates[,grep("Deg_",colnames(rates))])
)
```

```{r}
q <- quantile(gene.counts.dt[stau %in% c("STAU1", "PARIS")]$mean_count, probs = seq(0, 1, 0.05))
gene.counts.dt[stau %in% "None" & mean_count > q[2], stau := "None\n(Expression\nMatched)"]
gene.counts.dt[, .N, by = stau]

```

```{r}
rates.dt <- as.data.table(ratesInfo)
setnames(rates.dt, c("Syn", "Proc", "Deg"), c("Synthesis", "Processing", "Degradation"))
setkey(rates.dt, "Gene")
setkey(gene.counts.dt, "Gene")

# table(gene.counts.dt[stau != "None"]$Gene %in% rates.dt$Gene)
rates.dt <- rates.dt[gene.counts.dt]
rates.dt <- rates.dt[stau != "None"]

rates.melted.dt <- melt.data.table(rates.dt[, .(stau, Synthesis, Processing, Degradation)], id.vars = "stau")
rates.melted.dt$stau <- factor(rates.melted.dt$stau, levels = c("None", "None\n(Expression\nMatched)", "PARIS", "STAU1"))
# 
# wilcox.test(rates.dt[stau == "STAU1"]$Synthesis,
#             rates.dt[stau == "PARIS"]$Synthesis)
# 
# wilcox.test(rates.dt[stau == "STAU1"]$Processing,
#             rates.dt[stau == "PARIS"]$Processing)
# 
# wilcox.test(rates.dt[stau == "STAU1"]$Degradation,
#             rates.dt[stau == "PARIS"]$Degradation)

p <- ggplot(rates.melted.dt[value > 0], aes(x = stau, y = value)) +
  geom_violin(aes(fill = stau)) +
  geom_boxplot(width = 0.1, outlier.shape = NA) +
  scale_y_log10(label = comma) +
  annotation_logticks(sides = "l") +
  scale_fill_manual(values = gt_colours) +
  facet_wrap( ~ variable, scales = "free_y") +
  # geom_label(data = rates.dt[, .N, by = .(stau)], aes(label = N, y = 0), vjust = -0.1) +
  labs(x = "",
       y = "Rate",
       fill = "") +
  theme_minimal_grid() + theme(legend.position = "none")

if(save_plots) ggsave(p, filename = file.path(plot.path, "rna_metabolism_expression_matched.pdf"), width = 9, height = 6)

p
```

```{r}
wilcox.test(rates.dt[stau == "STAU1"]$Synthesis,
            rates.dt[stau == "None\n(Expression\nMatched)"]$Synthesis)

wilcox.test(rates.dt[stau == "STAU1"]$Processing,
            rates.dt[stau == "None\n(Expression\nMatched)"]$Processing)

wilcox.test(rates.dt[stau == "STAU1"]$Degradation,
            rates.dt[stau == "None\n(Expression\nMatched)"]$Degradation)
```

```{r}
p <- ggplot(rates.melted.dt[value > 0][stau != "PARIS"], aes(x = stau, y = value)) +
  geom_violin(aes(fill = stau)) +
  geom_boxplot(width = 0.1, outlier.shape = NA) +
  scale_y_log10(label = comma) +
  annotation_logticks(sides = "l") +
  scale_fill_manual(values = gt_colours) +
  facet_wrap( ~ variable, scales = "free_y") +
  # geom_label(data = rates.dt[, .N, by = .(stau)], aes(label = N, y = 0), vjust = -0.1) +
  labs(x = "",
       y = "Rate",
       fill = "") +
  theme_minimal_grid() + theme(legend.position = "none")

if(save_plots) ggsave(p, filename = file.path(plot.path, "rna_metabolism_expression_matched_ex_paris.pdf"), width = 8, height = 5)

p
```

## Reproduce with original gene list

```{r}
ys.dt <- as.data.table(readxl::read_xlsx(file.path(plot.path, "41586_2015_BFnature14280_MOESM50_ESM.xlsx"), skip = 1))
ys.genes <- unique(ys.dt[Annotation == "utr3"]$seqnames)

ys.rates.dt <- copy(rates.dt)
ys.rates.dt[, orig := ifelse(ensg %in% ys.dt$seqnames, "STAU1\nSugimoto", "none")]
ys.rates.dt[stau == "None\n(Expression\nMatched)", orig := "None\n(Expression\nMatched)"]
ys.rates.dt <- ys.rates.dt[orig != "none"]

ys.rates.melted.dt <- melt.data.table(ys.rates.dt[, .(orig, Synthesis, Processing, Degradation)], id.vars = "orig")
ys.rates.melted.dt$orig <- factor(ys.rates.melted.dt$orig, levels = c("None\n(Expression\nMatched)", "STAU1\nSugimoto"))

p <- ggplot(ys.rates.melted.dt[value > 0], aes(x = orig, y = value)) +
  geom_violin(aes(fill = orig)) +
  geom_boxplot(width = 0.1, outlier.shape = NA) +
  scale_y_log10(label = comma) +
  annotation_logticks(sides = "l") +
  scale_fill_tableau(palette = "Miller Stone") +
  facet_wrap( ~ variable, scales = "free_y") +
  # geom_label(data = rates.dt[, .N, by = .(stau)], aes(label = N, y = 0), vjust = -0.1) +
  labs(x = "",
       y = "Rate",
       fill = "") +
  theme_minimal_grid() + theme(legend.position = "none")

if(save_plots) ggsave(p, filename = file.path(plot.path, "rna_metabolism_ys.pdf"), width = 8, height = 5)
if(save_plots) ggsave(p, filename = file.path(plot.path, "rna_metabolism_ys.png"), width = 8, height = 5)

p
```

# RNA translation

```{r}
translation.dt <- fread("~/Dropbox (The Francis Crick)/comp_hiclip/4suseq/GSE84722_ST3.txt.gz")
setkey(translation.dt, Gene)
translation.dt <- translation.dt[gene.counts.dt]
translation.dt <- translation.dt[!is.na(CytNuc)]
translation.dt$stau <- factor(translation.dt$stau, levels = c("None", "None\n(Expression\nMatched)", "PARIS", "STAU1"))

p <- ggplot(translation.dt[stau != "None"], aes(x = stau, y = TrP)) +
  geom_violin(aes(fill = stau)) +
  geom_boxplot(width = 0.1, outlier.shape = NA) +
  coord_flip() +
  scale_fill_manual(values = gt_colours) +
  labs(x = "",
       y = "Translation potential",
       fill = "") +
  theme_minimal_grid() + theme(legend.position = "none")

if(save_plots) ggsave(p, filename = file.path(plot.path, "rna_translation.pdf"), width = 5, height = 4)
p
```
# RNA metabolism heatmap

```{r}
mat.dt <- rates.dt[stau %in% c("STAU1"), .(ensg, Synthesis, Processing, Degradation)]
mat.dt <- mat.dt[!is.na(Synthesis) | !is.na(Processing) | !is.na(Degradation)]
mat <- as.matrix(log10(mat.dt[, -1]))
# mat <- as.matrix(mat.dt[, -1])
rownames(mat) <- mat.dt$ensg

scaled.mat <- apply(mat, 2, scale)
rownames(scaled.mat) <- rownames(mat)
scaled.mat <- scaled.mat[!rowSums(is.na(scaled.mat)), ]
b <- max(abs(scaled.mat))

# PAM clustering
p <- factoextra::fviz_nbclust(scaled.mat, FUNcluster = cluster::pam, method = "silhouette", k.max = 5)
p + theme_minimal_grid()
```


```{r}
k.pam <- cluster::pam(scaled.mat, k = 3, metric = "euclidean", stand = TRUE)
annot.df <- data.frame(k.pam$clustering)

annot.df$cluster <- NA
annot.df[annot.df$k.pam.clustering == which.max(k.pam$medoids[, "Degradation"]), ]$cluster <- "A"
annot.df[annot.df$k.pam.clustering == which.min(k.pam$medoids[, "Degradation"]), ]$cluster <- "C"
annot.df[is.na(annot.df$cluster), ]$cluster <- "B"
annot.df$k.pam.clustering <- NULL

table(annot.df$cluster)
```

```{r}
stopifnot(all(rownames(annot.df) %in% rownames(scaled.mat))) # Ensure they are all there, before adjusting rows
annot.order <- rownames(annot.df)[order(annot.df$cluster, decreasing = FALSE)]
scaled.mat <- scaled.mat[match(annot.order, rownames(scaled.mat)), ]

if(save_plots) { 
  
  pheatmap(t(scaled.mat), 
         cluster_cols = FALSE, method = "complete", show_colnames = FALSE, scale = "none",
         breaks = unique(c(seq(-b, -2.5, length.out = 10), 
                           seq(-2.5, 2.5, length.out = 100), 
                           seq(2.5, b, length.out = 10))), 
         color = rev(colorRampPalette(brewer.pal(n = 11, name = "RdBu"))(120)),
         cluster_rows = FALSE,
         annotation_col = annot.df,
         annotation_colors = list("cluster" = metabolism_colours),
         filename = file.path(plot.path, "rna_metabolism_heatmap.pdf"),
         width = 10,
         height = 2)

}
```

```{r}
trp <- translation.dt[match(rownames(scaled.mat), translation.dt$ensg)]$TrP
scaled.mat <- cbind(scaled.mat, scale(trp))
colnames(scaled.mat)[4] <- "Translation"

if(save_plots) { 
  
  pheatmap(t(scaled.mat), 
           cluster_cols = FALSE, show_colnames = FALSE, scale = "none",
           breaks = unique(c(seq(-b, -2.5, length.out = 20),
                             seq(-2.5, 2.5, length.out = 80),
                             seq(2.5, b, length.out = 20))),
           color = rev(colorRampPalette(brewer.pal(n = 11, name = "RdBu"))(120)),
           cluster_rows = FALSE,
           annotation_col = annot.df,
           annotation_colors = list("cluster" = metabolism_colours),
           gaps_row = 3,
           filename = file.path(plot.path, "rna_metabolism_heatmap_with_translation.pdf"),
           width = 10,
           height = 2)

}

pheatmap(t(scaled.mat), 
         cluster_cols = FALSE, show_colnames = FALSE, scale = "none",
         breaks = unique(c(seq(-b, -2.5, length.out = 20),
                           seq(-2.5, 2.5, length.out = 80),
                           seq(2.5, b, length.out = 20))),
         color = rev(colorRampPalette(brewer.pal(n = 11, name = "RdBu"))(120)),
         cluster_rows = FALSE,
         annotation_col = annot.df,
         annotation_colors = list("cluster" = metabolism_colours),
         gaps_row = 3)
```

```{r}
utr3.duplexes.dt[ensg %in% rownames(annot.df)[annot.df$cluster == "A"], metabolism_cluster := "A"]
utr3.duplexes.dt[ensg %in% rownames(annot.df)[annot.df$cluster == "B"], metabolism_cluster := "B"]
utr3.duplexes.dt[ensg %in% rownames(annot.df)[annot.df$cluster == "C"], metabolism_cluster := "C"]
utr3.duplexes.dt <- utr3.duplexes.dt[!is.na(metabolism_cluster)]
utr3.duplexes.dt <- merge(utr3.duplexes.dt, mat.dt, by = "ensg")
```


## Integrate with Lugowski data

```{r}
lugowski.files <- list.files("~/Dropbox (The Francis Crick)/comp_hiclip/revisions/lugowski_et_al", full.names = TRUE)
lugowski.files <- grep("HEK.4SU.HL.[a-z]", lugowski.files, invert = TRUE, value = TRUE) # remove the exogenous as DRUID performed better with the same raw data
lugowski.list <- lapply(seq_along(lugowski.files), function(i) {
  
  dt <- fread(lugowski.files[i], select = c(1, 2))
  setnames(dt, c("HGNC", gsub(".csv.gz|GSE99517_", "", basename(lugowski.files[i]))))
  stopifnot(all(!duplicated(dt$HGNC)))
  return(dt)
  
})

lugowski.dt <- Reduce(function(x, y) merge(x, y, by = "HGNC", all.x = TRUE), lugowski.list)
```

```{r}
utr3.duplexes.dt[, HGNC := L_gene_name]
metabolism.dt <- merge(lugowski.dt, unique(utr3.duplexes.dt[, .(HGNC, L_gene_id, Degradation)]), by = "HGNC")
setnames(metabolism.dt, gsub("HEK\\.", "", names(metabolism.dt)))
```

### Correlation of rates

```{r}
lugowski.exps <- grep("one$|two$", colnames(metabolism.dt), value = TRUE)
p.list <- lapply(lugowski.exps, function(exp) {
  
  c <- cor(log10(metabolism.dt$Degradation), log10(1/metabolism.dt[, ..exp]), use = "pairwise.complete.obs", method = "spearman")
  p <- ggplot(metabolism.dt, aes(x = Degradation, y = 1/.data[[exp]])) +
        geom_point(alpha = 0.5, size = 1) +
        scale_x_log10() + scale_y_log10() +
        theme_minimal_grid() +
        annotate("text", x = Inf, y = Inf, hjust = 1, vjust = 1, colour = "steelblue4", label = paste0(round(c, 2)))
  
  return(p)
  
})

ggsave(file.path(plot.path, "lugowski.pdf"), wrap_plots(p.list, ncol = 2), width = 5, height = 7)
```

### Degradation heatmap

```{r}
metabolism.mat <- as.matrix(metabolism.dt[, grep("one$|two$", names(metabolism.dt)), with = FALSE])
rownames(metabolism.mat) <- tstrsplit(metabolism.dt$L_gene_id, "\\.")[[1]]

# degredation.mat <- metabolism.mat[match(rownames(metabolism.mat), rownames(scaled.mat)), ]

scaled.dt <- as.data.table(scaled.mat, keep.rownames = TRUE)
metabolism.dt[, rn := tstrsplit(metabolism.dt$L_gene_id, "\\.")[[1]]]
metabolism.dt[, L_gene_id := NULL]
metabolism.dt[, HGNC := NULL]
[, Degradation := NULL]
scaled.metabolism.dt <- metabolism.dt[, lapply(.SD, function(x) scale(log10(x))), .SDcols = -"rn"][, rn := metabolism.dt$rn]

all.scaled.dt <- merge(scaled.dt, scaled.metabolism.dt, by = "rn", all.x = TRUE)
all.scaled.mat <- as.matrix(all.scaled.dt[, -1])
rownames(all.scaled.mat) <- all.scaled.dt$rn
all.scaled.mat <- all.scaled.mat[match(rownames(scaled.mat), rownames(all.scaled.mat)), ]

degradation.mat <- all.scaled.mat[, grep("Synthesis|Processing|Translation", colnames(all.scaled.mat), invert = TRUE)]
colnames(degradation.mat) <- gsub("X|\\.V1", "", colnames(degradation.mat))
colnames(degradation.mat) <- gsub(".HL.", " rep", colnames(degradation.mat))
colnames(degradation.mat) <- gsub("repDRUID.", "DRUID rep", colnames(degradation.mat))
colnames(degradation.mat) <- gsub("one", "1", colnames(degradation.mat))
colnames(degradation.mat) <- gsub("two", "2", colnames(degradation.mat))
colnames(degradation.mat) <- gsub("Aman", "alpha-amantin", colnames(degradation.mat))
colnames(degradation.mat) <- gsub("ActD", "actinomycin D", colnames(degradation.mat))

pheatmap(degradation.mat, 
         cluster_cols = FALSE, show_colnames = TRUE, scale = "none",
         breaks = unique(c(seq(-b, -2.5, length.out = 20),
                           seq(-2.5, 2.5, length.out = 80),
                           seq(2.5, b, length.out = 20))),
         color = rev(colorRampPalette(brewer.pal(n = 11, name = "RdBu"))(120)),
         cluster_rows = FALSE, show_rownames = FALSE,
         annotation_row = annot.df,
         annotation_colors = list("cluster" = metabolism_colours),
         gaps_col = c(1, 3, 5, 7),
         gaps_row = cumsum(c(table(annot.df$cluster))),
         filename = file.path(plot.path, "rna_degradation_heatmap_with_lugowski.pdf"),
         width = 3,
         height = 10)
```

## Duplex span

```{r}
p <- ggplot(utr3.duplexes.dt[, mean(loop), by = .(metabolism_cluster, L_gene_id)], aes(x = V1, fill = metabolism_cluster)) +
  geom_density() +
  scale_fill_manual(values = metabolism_colours) +
  theme_minimal_grid() + theme(legend.position = "none") +
  facet_grid(. ~ metabolism_cluster) +
  scale_x_log10() + annotation_logticks(sides = "b") +
  labs(y = "Density",
       x = "Duplex span",
       fill = "")

if(save_plots) ggsave(p, filename = file.path(plot.path, "mean_duplex_span_density.pdf"), width = 8, height = 3)
p
```

## 3' UTR length

```{r}
kruskal.test(x = utr3.duplexes.dt[, mean(utr3_width), by = .(metabolism_cluster, L_gene_id)]$V1,
             g = utr3.duplexes.dt[, mean(utr3_width), by = .(metabolism_cluster, L_gene_id)]$metabolism_cluster)

p <- ggplot(utr3.duplexes.dt[, mean(utr3_width), by = .(metabolism_cluster, L_gene_id)], aes(x = metabolism_cluster, y = V1, fill = metabolism_cluster)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  coord_cartesian(ylim = c(0, 10000)) +
  scale_fill_manual(values = metabolism_colours) +
  theme_minimal_grid() + theme(legend.position = "none") +
  labs(x = "",
       y = "3' UTR length",
       fill = "")

if(save_plots) ggsave(p, filename = file.path(plot.path, "mean_utr3_length.pdf"), width = 4, height = 4)
p
```

## Circularisation score

```{r}
p <- ggplot(utr3.duplexes.dt[, mean(circ_score), by = .(metabolism_cluster, L_gene_id)], aes(x = metabolism_cluster, y = V1, fill = metabolism_cluster)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  coord_cartesian(ylim = c(0, 0.4)) +
  scale_fill_manual(values = metabolism_colours) +
  theme_minimal_grid() + theme(legend.position = "none") +
  labs(x = "",
       y = "Circularisation score",
       fill = "")

if(save_plots) ggsave(p, filename = file.path(plot.path, "mean_circ_score.pdf"), width = 4, height = 4)
p
```

```{r}
d <- dunn.test(x = utr3.duplexes.dt[, mean(circ_score), by = .(metabolism_cluster, L_gene_id)]$V1,
               g = utr3.duplexes.dt[, mean(circ_score), by = .(metabolism_cluster, L_gene_id)]$metabolism_cluster,
               method = "bh",
               alpha = 0.01)

d
```


## Relative 3' UTR positions

```{r}
longest.duplexes.dt <- utr3.duplexes.dt[, longest_loop := max(loop), by = ensg]
longest.duplexes.dt <- longest.duplexes.dt[loop == longest_loop]
stopifnot(any(duplicated(longest.duplexes.dt$ensg)))

longest.duplexes.dt[0 <= L_rel_start & L_rel_start < 0.33, L_rel_q := 1]
longest.duplexes.dt[0.33 <= L_rel_start & L_rel_start < 0.66, L_rel_q := 2]
longest.duplexes.dt[0.66 <= L_rel_start & L_rel_start <= 1, L_rel_q := 3]
longest.duplexes.dt[L_rel_start < 0, L_rel_q := 1]

longest.duplexes.dt[0 <= R_rel_start & R_rel_start < 0.33, R_rel_q := 1]
longest.duplexes.dt[0.33 <= R_rel_start & R_rel_start < 0.66, R_rel_q := 2]
longest.duplexes.dt[0.66 <= R_rel_start & R_rel_start <= 1, R_rel_q := 3]

longest.duplexes.dt[, rel_q := paste0(L_rel_q, "-", R_rel_q)]

longest.duplexes.dt$rel_q <- factor(longest.duplexes.dt$rel_q, 
                                       levels = c("1-1", "2-2", "3-3",
                                                  "1-2", "2-3",
                                                  "1-3"))

longest.duplexes.dt[rel_q %in% c("1-1", "2-2", "3-3"), q_span := "1 third"]
longest.duplexes.dt[rel_q %in% c("1-2", "2-3"), q_span := "2 thirds"]
longest.duplexes.dt[rel_q %in% c("1-3"), q_span := "3 thirds"]

p1 <- ggplot(longest.duplexes.dt, aes(x = q_span)) +
  geom_bar(width = 0.75, fill = "#4e79a7") +
  theme_minimal_grid() + theme(legend.position = "bottom") +
  labs(x = "",
       y = "Count",
       fill = "") +
  scale_y_log10()

p2 <- ggplot(longest.duplexes.dt, aes(x = q_span, fill = metabolism_cluster)) +
  geom_bar(position = "fill", width = 0.75) +
  scale_fill_manual(values = metabolism_colours) +
  theme_minimal_grid() + theme(legend.position = "bottom") +
  labs(x = "Duplex regions",
       y = "Fraction",
       fill = "")

if(save_plots) ggsave(plot_grid(p1, p2, ncol = 1, rel_heights = c(1, 2)), filename = file.path(plot.path, "thirds.pdf"), width = 4, height = 6)

plot_grid(p1, p2, ncol = 1, rel_heights = c(1, 2))
```