---
title: "Suppl. Fig 2"
author: "A. M. Chakrabarti"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_notebook: 
    code_folding: hide
    fig_align: center
    highlight: haddock
    theme: flatly
    toc: yes
    toc_depth: 4
    toc_float: yes
    fig_height: 7
    fig_width: 7
---

# Libraries and functions

```{r}
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(toscatools))
suppressPackageStartupMessages(library(rtracklayer))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(scales))
suppressPackageStartupMessages(library(ggthemes))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(UpSetR))
suppressPackageStartupMessages(library(tidyverse))

plot.path <- "~/Dropbox (The Francis Crick)/comp_hiclip/revisions/plots/figure_2"
if(!dir.exists(plot.path)) dir.create(plot.path, recursive = TRUE)
ht_colours <- c("Linker" = "#8175aa", "Direct" = "#6fb899", "Short-range" = "#31a1b3")
```


```{r}
convert_linkerbam_to_hybrids <- function(bam, threads = 4, export_fasta = FALSE) {

  ga <- GenomicAlignments::readGAlignments(bam,
                                           use.names = TRUE,
                                           param = Rsamtools::ScanBamParam(what = c("seq", "qual")))

  ga <- ga[GenomicAlignments::njunc(ga) == 0]

  cig <- GenomicAlignments::cigarRangesAlongQuerySpace(GenomicAlignments::cigar(ga),
                                                       after.soft.clipping = TRUE,
                                                       drop.empty.ranges = TRUE,
                                                       reduce.ranges = TRUE)
  stopifnot(all(S4Vectors::elementNROWS(cig) == 1))
  cig <- unlist(IRanges::IRangesList(cig))

  # Probably a vectorised way of doing this but seq[cig] doesn't work?
  S4Vectors::mcols(ga)$seq <- Biostrings::DNAStringSet(parallel::mclapply(seq_along(cig), function(i) { S4Vectors::mcols(ga)$seq[[i]][cig[i]] }, mc.cores = threads))
  S4Vectors::mcols(ga)$qual <- Biostrings::BStringSet(parallel::mclapply(seq_along(cig), function(i) { S4Vectors::mcols(ga)$qual[[i]][cig[i]] }, mc.cores = threads))

  dt <- as.data.table(ga, keep.rownames = TRUE)
  dt[, `:=` (name = gsub("^L\\.|^R\\.", "", rn),
             arm = tstrsplit(rn, "\\.")[[1]])][, rn := NULL]

  hybrids.dt <- merge(dt[arm == "L"], dt[arm == "R"], by = "name")
  hybrids.dt[, seq := paste0(seq.x, seq.y)]
  hybrids.dt[, seq_width := nchar(seq)]

  if(export_fasta) {

    hybrids.fa <- Biostrings::DNAStringSet(paste0(hybrids.dt$seq.x, hybrids.dt$seq.y))
    names(hybrids.fa) <- hybrids.dt$name

    hybrids.qual <- Biostrings::BStringSet(paste0(hybrids.dt$qual.x, hybrids.dt$qual.y))


    Biostrings::writeXStringSet(hybrids.fa, export_fasta,
                                format = "fastq",
                                qualities = hybrids.qual,
                                compress = TRUE)

  }

  return(hybrids.dt)

}
```

# In silico data generation

## Load linker hybrids data

```{r}
linker.path <- "/Volumes/lab-luscomben/home/shared/projects/ira-nobby/comp_hiclip/revisions/results_linker"

linker.dt <- rbindlist(list(fread(file.path(linker.path, "stau1_high.hybrids.gc.annotated.tsv.gz"))[, sample := "stau1_high"],
                            fread(file.path(linker.path, "stau1_low.hybrids.gc.annotated.tsv.gz"))[, sample := "stau1_low"]),
                       use.names = TRUE)
```

## Generate in silico no linker reads

```{r}
silico.path <- "/Volumes/lab-luscomben/home/shared/projects/ira-nobby/comp_hiclip/revisions/silico"
if(!dir.exists(silico.path)) dir.create(silico.path, recursive = TRUE)

bam.files <- list.files(linker.path, pattern = ".bam$", full.names = TRUE)

lapply(bam.files, function(x) {
  
  exp <- tstrsplit(basename(x), "\\.")[[1]]
  message(exp)
  
  bam.hybrids.dt <- convert_linkerbam_to_hybrids(x, threads = 4)
  bam.hybrids.dt <- bam.hybrids.dt[name %in% linker.dt[sample == exp]$name]
  stopifnot(all(linker.dt[sample == exp]$name %in% bam.hybrids.dt$name))
  
  bam.fa <- Biostrings::DNAStringSet(paste0(bam.hybrids.dt$seq.x, bam.hybrids.dt$seq.y))
  names(bam.fa) <- bam.hybrids.dt$name
  bam.qual <- Biostrings::BStringSet(paste0(bam.hybrids.dt$qual.x, bam.hybrids.dt$qual.y))
  
  Biostrings::writeXStringSet(bam.fa, file.path(silico.path, paste0(exp, ".silico.fastq.gz")),
                              format = "fastq",
                              qualities = bam.qual,
                              compress = TRUE)
  
})
```

## Generate sample sheet

```{r}
samplesheet.dt <- data.table(fastq = list.files(silico.path, full.names = TRUE, pattern = "fastq.gz$"))
samplesheet.dt[, sample := tstrsplit(basename(fastq), "\\.")[[1]]]
samplesheet.dt[, fastq := gsub("/Volumes/lab-luscomben/", "/camp/lab/luscomben/", fastq)]
setcolorder(samplesheet.dt, c("sample", "fastq"))
fwrite(samplesheet.dt, file.path(silico.path, "silico.csv"))
```

## Run Tosca

```{bash, eval = FALSE}
#!/bin/bash

ml purge
ml Nextflow/21.10.3
ml Singularity/3.6.4
ml Graphviz/2.38.0-foss-2016b

export NXF_SINGULARITY_CACHEDIR=/camp/lab/luscomben/home/shared/singularity

REFDIR=/camp/lab/luscomben/home/shared/projects/ira-nobby/comp_hiclip/ref

cd /camp/lab/luscomben/home/shared/projects/ira-nobby/comp_hiclip/revisions/silico

nextflow pull amchakra/tosca -r v1.0.0

nextflow run amchakra/tosca -r v1.0.0 \
-resume \
-profile crick \
--input silico.csv \
--outdir results_silico \
--genome_fai $REFDIR/GRCh38.primary_assembly.genome.fa.fai \
--star_genome $REFDIR/STAR_GRCh38_GencodeV33 \
--transcript_fa $REFDIR/GRCh38.gencode_v33.fa \
--transcript_fai $REFDIR/GRCh38.gencode_v33.fa.fai \
--transcript_gtf $REFDIR/GRCh38.gencode_v33.tx.gtf.gz \
--regions_gtf $REFDIR/icount_mini_utr3/regions.gtf.gz \
--percent_overlap 0.5
```

# Results

```{r}
linker.path <- "/Volumes/lab-luscomben/home/shared/projects/ira-nobby/comp_hiclip/revisions/results_linker"
linker.dt <- rbindlist(list(fread(file.path(linker.path, "stau1_high.hybrids.gc.annotated.tsv.gz"))[, sample := "stau1_high"],
                            fread(file.path(linker.path, "stau1_low.hybrids.gc.annotated.tsv.gz"))[, sample := "stau1_low"]),
                       use.names = TRUE)

results.path <- "/Volumes/lab-luscomben/home/shared/projects/ira-nobby/comp_hiclip/revisions/silico/results_silico"
silico.dt <- rbindlist(list(fread(file.path(results.path, "hybrids", "stau1_high.hybrids.gc.annotated.tsv.gz"))[, sample := "stau1_high"],
                            fread(file.path(results.path, "hybrids", "stau1_low.hybrids.gc.annotated.tsv.gz"))[, sample := "stau1_low"]),
                       use.names = TRUE)

linker.dt[, recovered := ifelse(name %in% silico.dt$name, "Recovered", "Not recovered")]
linker.dt[, readlength := L_width + R_width]
linker.dt[, minwidth := min(L_width, R_width), by = name]
linker.dt[, type := ifelse(L_seqnames == R_seqnames, "intra-transcript", "inter-transcript")]
```

# Suppl. Fig 2 A - Pie

```{r}
pie.dt <- linker.dt[, .N, by = .(recovered)]
pie.dt[, percent := paste0(round(N/sum(pie.dt$N) * 100, 1), "%")]
pie.dt[, lab.ypos := cumsum(N) - 0.5 * N]

p <- ggplot(pie.dt, aes(x = "", y = N, fill = recovered)) + 
  geom_bar(stat = "identity") + 
  coord_polar("y", start = 1.571) +
  scale_fill_tableau(palette = "Classic Blue-Red 6", direction = -1) +
  theme_minimal_grid() + theme(legend.position = "top",
                               axis.text = element_blank(),
                               axis.ticks = element_blank(),
                               axis.line = element_blank(),
                               panel.grid.major = element_blank()) +
  labs(x = "",
       y = "",
       fill = "") +
  geom_text(aes(y = lab.ypos, label = paste0(N, "\n", percent)), color = "white", size = 5)

ggsave(p, filename = file.path(plot.path, "silico_recovery.pdf"), width = 7, height = 4)

p
```

# Suppl. Fig 2 A - Matches

```{r}
comp.dt <- merge(linker.dt, silico.dt, by = "name")
comp.dt[, L_match := ifelse(L_seqnames.x != L_seqnames.y, FALSE, TRUE)]
comp.dt[, R_match := ifelse(R_seqnames.x != R_seqnames.y, FALSE, TRUE)]
comp.dt[, L_start_match := ifelse(L_start.x >= L_start.y - 10 & L_start.x <= L_start.y + 10, TRUE, FALSE)]
comp.dt[, R_start_match := ifelse(R_start.x >= R_start.y - 10 & R_start.x <= R_start.y + 10, TRUE, FALSE)]

match.dt <- comp.dt[, .(L_match, R_match, L_start_match, R_start_match)]
match.dt[L_match == TRUE & L_start_match == TRUE | R_match == TRUE & R_start_match == TRUE, match := "One arm matches"]
match.dt[L_match == TRUE & R_match == TRUE & L_start_match == TRUE & R_start_match == TRUE, match := "Both arms match"]
match.dt[is.na(match), match := "No match"]

match.dt$match <- factor(match.dt$match, levels = rev(c("No match", "One arm matches", "Both arms match")))
match.dt <- match.dt[, .N, by = match]
match.dt[match == "Both arms match", percent := paste0(round(N/sum(match.dt$N) * 100, 1), "%")]
match.dt[is.na(percent), percent := ""]

p <- ggplot(match.dt, aes(x = "", y = N, fill = match)) +
  geom_col(position = "fill") + coord_flip() +
  scale_y_continuous(label = percent) +
  scale_fill_manual(values = c("No match" = "#ea6b73",
                               "One arm matches" = "#6ba3d6",
                               "Both arms match" = "#2c69b0"),
                    guide = guide_legend(reverse = TRUE)) +
  theme_minimal_grid() + theme(legend.position = "bottom") +
  labs(x = "",
       y = "Percentage",
       fill = "") +
  geom_text(aes(label = percent), y = 0.5, color = "white", size = 5)

ggsave(p, filename = file.path(plot.path, "silico_recovery_match.pdf"), width = 7, height = 2)

p
```

# Suppl. Fig 2 B - Read length

```{r}
p <- ggplot(linker.dt, aes(x = readlength, colour = recovered)) +
  geom_density() +
  facet_grid(sample ~ ., 
             labeller = labeller(sample = c(stau1_high = "High RNase", stau1_low = "Low RNase"))) +
  scale_colour_tableau(palette = "Classic Blue-Red 6", direction = -1) +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "Read length",
       y = "Density",
       colour = "")

ggsave(p, filename = file.path(plot.path, "silico_recovery_readlength.pdf"), width = 7, height = 4)

p
```

```{r}
linker.dt[, median(readlength), by = recovered]
wilcox.test(x = linker.dt[recovered == "Recovered"]$readlength, y = linker.dt[recovered == "Not recovered"]$readlength)
```

# Suppl. Fig 2 C - shortest arm

```{r}
p <- ggplot(linker.dt, aes(y = minwidth, x = recovered, colour = recovered)) +
  geom_violin() +
  geom_boxplot(width = 0.1, outlier.shape = NA) +
  coord_flip() +
  facet_grid(sample ~ ., 
           labeller = labeller(sample = c(stau1_high = "High RNase", stau1_low = "Low RNase"))) +
  scale_colour_tableau(palette = "Classic Blue-Red 6", direction = -1) +
  theme_minimal_grid() + theme(legend.position = "none") +
  labs(x = "",
       y = "Shorter arm length",
       colour = "Recovered")

ggsave(p, filename = file.path(plot.path, "silico_recovery_minarmlength.pdf"), width = 7, height = 4)
p
```

```{r}
linker.dt[, median(minwidth), by = recovered]
wilcox.test(x = linker.dt[recovered == "Recovered"]$minwidth, y = linker.dt[recovered == "Not recovered"]$minwidth)
```

# Suppl. Fig 2 D - inter-intra

```{r}
p <- linker.dt %>%
  group_by(recovered, type) %>%
  summarise(reads = n()) %>%
  mutate(percentage = reads/sum(reads)) %>%
  ggplot(aes(x = recovered, y = percentage, fill = type)) +
  geom_col(width = 0.5, position = "fill") +
  geom_text(aes(label = ifelse(percentage > 0.05, paste0(round(percentage * 100, 1), "%"), "")), 
            position = position_stack(vjust=0.5),
            colour = "white",
            size = 5) +
  coord_flip() + 
  scale_y_continuous(label = percent) +
  scale_fill_tableau(palette = "Seattle Grays", guide = guide_legend(reverse = TRUE)) +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "",
       y = "Percentage of hybrids",
       fill = "")

ggsave(p, filename = file.path(plot.path, "silico_recovery_interintra.pdf"), width = 7, height = 4)

p
```

# Suppl. Fig 2 E - Regions

```{r}
arms.dt <- rbindlist(list(linker.dt[, .(sample, recovered, L_region)], linker.dt[, .(sample, recovered, R_region)]), use.names = FALSE)
arms.dt$L_region <- factor(arms.dt$L_region, levels = rev(c("rRNA", "tRNA", "UTR5", "CDS", "intron", "UTR3", "ncRNA")))

p <- arms.dt %>%
  group_by(L_region, recovered) %>%
  summarise(arms = n()) %>%
  group_by(recovered) %>%
  mutate(percentage = arms/sum(arms)) %>%
  ggplot(aes(x = recovered, y = percentage, fill = L_region)) +
  geom_col(width = 0.5, position = "fill") +
  geom_text(aes(label = ifelse(percentage > 0.05, paste0(round(percentage * 100, 1), "%"), "")), 
            position = position_stack(vjust=0.5),
            colour = "white",
            size = 5) +
  coord_flip() +
  scale_y_continuous(label = percent) +
  scale_fill_tableau(palette = "Miller Stone", 
                     direction = -1, 
                     guide = guide_legend(reverse = TRUE)) +
  theme_minimal_grid() + 
  theme(legend.position = "top") +
  labs(x = "",
       y = "Percentage of hybrids",
       fill = "")

ggsave(p, filename = file.path(plot.path, "silico_recovery_regions.pdf"), width = 7, height = 4)

p
```

# Session info

```{r}
session_info()
```


